### PR TITLE
Separate persistent mutable state from config

### PR DESCRIPTION
**Pull request by matklad**
That way, we clearly see which things are not change, and we also
clearly see which things are persistent.

r? @Veetaha

### CODE REVIEW

File: editors/code/src/main.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { ensureServerBinary } from './installation/server';
 import { Config } from './config';
 import { log } from './util';
 import { ensureProperExtensionVersion } from './installation/extension';
+import { PersistentState } from './persistentState';
```
Comment (Veetaha) [inline editors/code/src/main.ts]: @matklad, I'd be consistent in file name case conventions...

**Comment by Veetaha** [comment]
LGTM modulo the most important thing

**Comment by matklad** [comment]
bors r+

We should add a tidy test in Rust to check casing of `.ts` file. It's better to roll our own code for this than to use esling-filenames-plugin.

**Comment by Veetaha** [comment]
What is the rationale not to use the `eslint` plugin anyway?

**Comment by matklad** [comment]
If eslint lint folks don’t think that this is trustworthy/important enough to bring in core, than I don’t think using a plugin would be a net benefit over writing a dozen of lines of rust code.

**Comment by matklad** [comment]
In general, the guideline I use is “don’t add a dependency unless you have to”.

**Comment by Veetaha** [comment]
But this is a dev dependency and the code is already written for us. This is actually the opposite since the eslint guys thought that we would not want all the rules at once, so that we only include the ones we are interested in as dependencies...

Also notice, this is not me who elaborated on using a dependency here at all ;P

**Comment by Veetaha** [comment]
Going further, we should not use `eslint` at all, since its plugins are the core idea of it...

**Comment by Veetaha** [comment]
The bors is stuck?...

**Comment by matklad** [comment]
So, I've actually timed myself, and implementing the check for file names in rust took me 5 minutes, 20 seconds (of mostly coding). Adding the type script plugin took me 3 minutes 50 seconds (of mostly reading the docs). Indeed there's a savings of about 1.5 minutes of initial implementation work if I go with dependency and not re-implementation. 

The lint implementation contains this line:

```json
  "filenames/match-regex": [2, "^[a-z_]+$", true]
```

I understand what `"^[a-z_]+$"` does, but I have no idea what 2 or `true` means. I've also haven't read the code of the dependency, so I need to trust it (and my trust level is slow, as I am not well-versed in nodejs ecosystem, and the repository with the plugin does not seem high profile). 

This is in contrast to rust implementation, which is obvious to me, can be easily tweaked (and, indeed, I've already though that it might be extended to cover no-FIXME check as well) and doesn't add transitive dependencies on `"lodash.camelcase": "4.3.0"`, `"lodash.kebabcase": "4.1.1",`, ` "lodash.snakecase": "4.1.1"` and `"lodash.upperfirst": "4.3.1"`. 


I do believe that dependencies have a non-trivial hidden costs, *especially* if they are dependencies from an ecosystem where I can't judge if a dependency is reliable or not. I also think that micro-dependencies and ultra-configurability for *tooling* is just wrong, and probably creates more maintenance burden than it saves by enforcing consistency. I have high hopes for the [rome](https://github.com/facebookexperimental/rome) project which seems like it could, in the future, provide a single box which I can just use without too much googling around.

**Comment by Veetaha** [comment]
Those `lodash.*` packages are single-function-modules (they were created because people don't want to pull the whole `lodash` library into their javascript bundles), and this is a not a bad idea!...

The tools that are not frameworks (that dictate you how to write code) do have to be ultra-configurable to support as many use cases as possible, but they always provide reasonable defaults, such that ultra-configurability is opt-in and not opt-out.

I tend to believe the ecosystem (at least 500K+ downloads packages), this is what reliefs you from maintaining your own impl and if you don't like existing one, you can always build your own from smaller pieces.

I see you just want some framework that will bundle all the necessary tools you will need. I do like when everything is consistent, when each module uses the same conventions and interface as the other one (e.g. Angular, Unreal Engine, Qt are the tools you would use and very rarely demand any ad hoc ultra-configurable dependency) Unfortunately `vscode` doesn't provide the whole framework for our use case...

**Comment by matklad** [comment]
I don't feel like arguing about this more :) I propose to agree to disagree on our subjective assessment of cost/benefit ratio of external dependencies.

**Pull request merged**